src/metabase/models/pulse.clj

```diff
@@ -0,0 +1,159 @@
+                :creator  (delay (when creator_id (db/sel :one User :id creator_id))))))
+
+  (pre-cascade-delete [_ {:keys [id]}]
+    (db/cascade-delete PulseCard :pulse_id id)
+    (db/cascade-delete PulseChannel :pulse_id id)))
+
+(extend-ICanReadWrite PulseEntity :read :public-perms, :write :public-perms)
+
+
+;; ## Helper Functions
+
+(defn- update-pulse-cards
+  "Update the `PulseCards` for a given PULSE.
+   CARD-IDS should be a definitive collection of *all* IDs of cards for the pulse in the desired order.
+
+   *  If an ID in CARD-IDS has no corresponding existing `PulseCard` object, one will be created.
+   *  If an existing `PulseCard` has no corresponding ID in CARD-IDs, it will be deleted.
+   *  All cards will be updated with a `position` according to their place in the collection of CARD-IDS"
+  {:arglists '([pulse card-ids])}
+  [{:keys [id]} card-ids]
+  {:pre [(integer? id)
+         (coll? card-ids)
+         (every? integer? card-ids)]}
+  ;; first off, just delete any cards associated with this pulse (we add them again below)
+  (k/delete PulseCard (k/where {:pulse_id id}))
+  ;; now just insert all of the cards that were given to us
+  (let [cards (map-indexed (fn [idx itm] {:pulse_id id :card_id itm :position idx}) card-ids)]
+    (k/insert PulseCard (k/values cards))))
+
+(defn- create-update-delete-channel
+  "Utility function which determines how to properly update a single pulse channel."
+  [pulse-id new-channel existing-channel]
+  ;; NOTE that we force the :id of the channel being updated to the :id we *know* from our
+  ;;      existing list of `PulseChannels` pulled from the db to ensure we affect the right record
+  (let [channel (when new-channel (assoc new-channel :pulse_id pulse-id :id (:id existing-channel)))]
+    (cond
+      ;; 1. in channels, NOT in db-channels = CREATE
+      (and channel (not existing-channel))  (pulse-channel/create-pulse-channel channel)
+      ;; 2. NOT in channels, in db-channels = DELETE
+      (and (nil? channel) existing-channel) (db/cascade-delete PulseChannel :id (:id existing-channel))
+      ;; 3. in channels, in db-channels = UPDATE
+      (and channel existing-channel)        (pulse-channel/update-pulse-channel channel)
+      ;; 4. NOT in channels, NOT in db-channels = NO-OP
+      :else nil)))
+
+(defn- update-pulse-channels
+  "Update the `PulseChannels` for a given PULSE.
+   CHANNELS should be a definitive collection of *all* of the channels for the the pulse.
+
+   * If a channel in the list has no existing `PulseChannel` object, one will be created.
+   * If an existing `PulseChannel` has no corresponding entry in CHANNELS, it will be deleted.
+   * All previously existing channels will be updated with their most recent information."
+  {:arglists '([pulse channels])}
+  [{:keys [id]} channels]
+  {:pre [(integer? id)
+         (coll? channels)
+         (every? map? channels)]}
+  (let [new-channels   (group-by #(keyword (:channel_type %)) channels)
```

Comment:
Instead of `#(keyword (:channel_type %))` you can do `(comp keyword :channel_type)` here. `comp` "composes" functions and returns a new one that applies them in order from right to left